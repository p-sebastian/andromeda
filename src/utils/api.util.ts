import { ServerEnum } from './enums.util'
import { TServerState } from '@reducers/server.reducer'
import { memoize, values } from 'lodash'
import { Observable, OperatorFunction, of, iif } from 'rxjs'
import { ajax } from 'rxjs/ajax'
import { RootState } from '@reducers/index'
import { StateObservable } from 'redux-observable'
import { withLatestFrom, map, mergeMap, tap, delay } from 'rxjs/operators'
import { AjaxResponse } from 'rxjs/internal/observable/dom/AjaxObservable'
import { ApiActionsType } from '@actions/index'
import { isOfType } from 'typesafe-actions'
import { logger } from './logger.util'
import { TActions } from './types.util'
import { nrmlz } from './normalizr.util'

type Method = 'GET' | 'POST' | 'PUT' | 'DELETE'
type AjaxCreator = (
  endpoint: any,
  method: Method,
  body: any
) => Observable<AjaxResponse>
/**
 * @param usable params that will be used to create the key for caching
 */
const _ajaxConfig = ({
  enabled,
  status,
  ...usable
}: TServerState): AjaxCreator => (endpoint, method, body) => {
  const memo = memoize(formUrl, args => values(args).join('_'))
  const [baseURL, headers] = memo(usable as any)
  let url = `${baseURL}${endpoint}`
  if (method === 'GET') {
    url += `?${queryString(body)}`
  }
  logger.log(url)
  // sending anything in the body on a GET
  // breaks the request
  return ajax({
    headers,
    method,
    url,
    body: method === 'GET' ? undefined : body,
    timeout: 5000
  })
}
type WithApi = (
  state$: StateObservable<RootState>,
  method: Method
) => OperatorFunction<ApiActionsType, Observable<[ApiActionsType, any]>>
/**
 * returns mapped tuple [action, responseObservable]
 * all actions merging this MUST HAVE meta with ServerEnum
 * @param state$ passed when calling manually
 * @param method: {Method}
 * @arg action$ incoming input stream
 */
export const withApi: WithApi = (state$, method) => action$ =>
  action$.pipe(
    withLatestFrom(state$),
    map(([action, state]) => {
      const { endpoint, params } = action.payload
      return _ajaxConfig(state.server[action.meta])(
        endpoint,
        method,
        params
      ).pipe(
        tap(res => logger.log('AJAX_RESPONSE', res)),
        delay(2000),
        map(res => [action, res.response] as [ApiActionsType, any])
      )
    })
  )
/**
 * [C] type constant
 * [A] incoming action
 */
type OnCase<C = string, A = TActions> = (
  CONSTANT: C
) => (callback: (response: any) => A) => OperatorFunction<[A, any], [A, any]>
/**
 *
 * @param CONSTANT Action constant
 */
export const onCase: OnCase = CONSTANT => callback => action$ =>
  action$.pipe(
    mergeMap(
      ([action, response]) =>
        // iif must return an Observable
        iif(
          () => isOfType(CONSTANT, action),
          of([callback(nrmlz(CONSTANT, response)), response]),
          of([action, response])
        ) as Observable<[TActions, any]>
    )
    // tap(([, res]) => logger.info('normalized', res))
  )

type FormUrl = (
  server: TServerState,
  ssl?: boolean
) => [string, { [key: string]: string }]
/**
 * returns tuple [url, headers]
 * @param server
 * @param ssl use https
 */
const formUrl: FormUrl = (
  { key, lanPort, lanUrl, remotePort, remoteUrl, endpoint, apiKey },
  ssl = false
) => {
  const url = endpoint === 'lan' ? lanUrl : remoteUrl
  const port = endpoint === 'lan' ? lanPort : remotePort

  switch (key) {
    case ServerEnum.SONARR:
      return [
        `http${ssl ? 's' : ''}://${url}:${port}/api`,
        { 'X-Api-Key': apiKey }
      ]
    default:
      return ['', { '0': '0' }]
  }
}

const queryString = (body: any = {}) =>
  Object.keys(body)
    .map(key => `${key}=${body[key]}`)
    .join('&')
