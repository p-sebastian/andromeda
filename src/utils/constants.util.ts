import { ThemeEnum, ServerEnum, ColorEnum, GradientEnum } from './enums.util'
import { TAvailableServers, TGradientConstants } from './types.util'

export const AVAILABLE_SERVERS: TAvailableServers = Object.freeze({
  [ServerEnum.SONARR]: {
    key: ServerEnum.SONARR,
    title: 'sonarr',
    themeKey: ThemeEnum.SONARR,
    tabs: ['upcoming', 'shows']
  },
  [ServerEnum.RADARR]: {
    key: ServerEnum.RADARR,
    title: 'radarr',
    themeKey: ThemeEnum.RADARR,
    tabs: []
  },
  [ServerEnum.LIDARR]: {
    key: ServerEnum.LIDARR,
    title: 'lidarr',
    themeKey: ThemeEnum.LIDARR,
    tabs: []
  },
  [ServerEnum.SABNZBD]: {
    key: ServerEnum.SABNZBD,
    title: 'sabnzbd',
    themeKey: ThemeEnum.SABNZBD,
    tabs: []
  },
  [ServerEnum.TORRENT]: {
    key: ServerEnum.TORRENT,
    title: 'torrent',
    themeKey: ThemeEnum.TORRENT,
    tabs: []
  }
})
export const COLORS = Object.freeze({
  [ColorEnum.MAIN]: 'hsla(225, 6%, 13%, 1)',
  [ColorEnum.SONARR]: 'hsla(195, 93%, 39%, 1)',
  [ColorEnum.RADARR]: 'hsla(42, 99%, 53%, 1)',
  [ColorEnum.LIDARR]: 'hsl(154, 100%, 29%)',
  [ColorEnum.SABNZBD]: 'hsl(45, 94%, 49%)',
  [ColorEnum.TORRENT]: 'hsl(0, 100%, 29%)',
  [ColorEnum.GRAY]: 'hsla(228, 11%, 28%, 1)',
  [ColorEnum.PLACEHOLDER]: 'hsla(0, 0%, 52%, 1)',
  [ColorEnum.DANGER]: '#F44336',
  [ColorEnum.SUCCESS]: '#1DE9B6',
  [ColorEnum.INFO]: '#00B0FF'
})
export const SERVER_GRADIENTS = Object.freeze({
  [ServerEnum.SONARR]: 'hsla(195, 93%, 39%, 1)',
  [ServerEnum.RADARR]: 'hsla(42, 99%, 53%, 1)',
  [ServerEnum.LIDARR]: 'hsl(154, 100%, 29%)',
  [ServerEnum.SABNZBD]: 'hsl(45, 94%, 49%)',
  [ServerEnum.TORRENT]: 'hsl(0, 100%, 29%)'
})
export const GRADIENTS: TGradientConstants = Object.freeze({
  [GradientEnum.ORANGE]: {
    colors: ['#C6FFDD', '#FBD786', '#f7797d'],
    start: [0, 0.5] as [number, number],
    end: [1.3, 0.5] as [number, number]
  },
  [GradientEnum.GREEN]: {
    colors: ['#34e89e', '#218b6f', '#144f50', '#0f3443'],
    start: [0.2, 0] as [number, number],
    end: [0.9, 1] as [number, number]
  },
  [GradientEnum.BLUE]: {
    colors: ['#12c2e9', '#c471ed', '#f7797d'],
    start: [0, 0.5] as [number, number],
    end: [1, 0.5] as [number, number]
  },
  [GradientEnum.PURPLE]: {
    colors: ['#1096e1', '#6651e5', '#b214e8'],
    start: [0.3, 0] as [number, number],
    end: [0.6, 1] as [number, number]
  },
  [GradientEnum.RED]: {
    colors: ['#ee0979', '#ff6a00'],
    start: [0, 0.5] as [number, number],
    end: [1, 0.5] as [number, number]
  },
  [GradientEnum.BUTTONS]: {
    colors: ['#B2FEFA', '#0ed2f7'],
    start: [0, 0.5] as [number, number],
    end: [1, 0.5] as [number, number]
  }
})
