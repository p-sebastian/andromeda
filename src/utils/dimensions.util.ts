import { Dimensions } from 'react-native'

export const SCREEN_HEIGHT = Dimensions.get('window').height
export const SCREEN_WIDTH = Dimensions.get('window').width
// Sidebar size
export const OFFSET = (SCREEN_WIDTH * 10) / 100
export const DRAWER_WIDTH = (SCREEN_WIDTH * 75) / 100 + OFFSET
export const HIDDEN_WIDTH = DRAWER_WIDTH - OFFSET
/**
 * Width between the inner and outer bar
 * of the side menu
 */
export const BETWEEN_WIDTH = DRAWER_WIDTH - OFFSET * 3
