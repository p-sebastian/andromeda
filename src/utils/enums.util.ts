export enum ThemeEnum {
  MAIN = 1,
  SONARR,
  RADARR,
  LIDARR,
  SABNZBD,
  TORRENT
}

export enum ColorEnum {
  MAIN = 1,
  SONARR,
  RADARR,
  LIDARR,
  SABNZBD,
  TORRENT,
  GRAY = 500,
  PLACEHOLDER,
  DANGER,
  SUCCESS,
  INFO
}
export enum GradientEnum {
  BLUE = 1,
  GREEN,
  ORANGE,
  PURPLE,
  RED,
  BUTTONS,
  SEASONS
}

export enum ServerEnum {
  SONARR = 'sonarr',
  RADARR = 'radarr',
  LIDARR = 'lidarr',
  SABNZBD = 'sabnzbd',
  TORRENT = 'torrent'
}
