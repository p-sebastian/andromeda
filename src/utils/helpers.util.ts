import { ScreenNames } from '@routes'
import { AVAILABLE_SERVERS } from './constants.util'
import Fuse from 'fuse.js'
import { SCREEN_HEIGHT, SCREEN_WIDTH } from '@utils/dimensions.util'
import { Platform } from 'react-native'
import { memoize } from 'lodash'
import { TServer } from '@utils/types.util'

const _withinScreen = (screenName: ScreenNames) => {
  let _server = null
  Object.values(AVAILABLE_SERVERS).forEach(server => {
    if (
      server.title === screenName ||
      server.tabs.findIndex(n => n === screenName) > -1
    ) {
      _server = { ...server }
    }
  })
  return _server as TServer | null
}
export const withinScreen = memoize(_withinScreen)

export const byteToGB = (size: number = 0) => {
  const gb = 1073741824
  return (size / gb).toFixed(1)
}

const fuseOptions = {
  keys: ['title'],
  id: 'id'
}
export const fuzzySearch = (data: {}) =>
  new Fuse(Object.values(data), fuseOptions)

export const isIphoneX = () =>
  (Platform.OS === 'ios' &&
    !Platform.isPad &&
    !Platform.isTVOS &&
    (SCREEN_HEIGHT === 812 || SCREEN_WIDTH === 812)) ||
  (SCREEN_HEIGHT === 896 || SCREEN_WIDTH === 896)
