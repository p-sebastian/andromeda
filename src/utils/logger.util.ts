import Reactotron from 'reactotron-react-native'

export const logger = {
  log(...args: any[]) {
    Reactotron.log!(args)
  },
  info(...args: any[]) {
    Reactotron.logImportant!(args)
  },
  warn(...args: any[]) {
    Reactotron.warn!(args)
  },
  error(...args: any[]) {
    Reactotron.error!(args, '')
  }
}
