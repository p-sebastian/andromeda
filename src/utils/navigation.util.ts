import {
  NavigationActions,
  StackActions,
  NavigationContainerComponent
} from 'react-navigation'

export class RootNavigation {
  private static _instance: RootNavigation
  private _navigator: NavigationContainerComponent | undefined

  static get instance() {
    console.info('called')
    if (!RootNavigation._instance) {
      RootNavigation._instance = new RootNavigation()
    }
    return RootNavigation._instance
  }

  set navigator(ref: NavigationContainerComponent) {
    console.info('did set')
    this._navigator = ref
    RootNavigation._instance = new RootNavigation()
    if (!this._navigator) {
    }
  }

  /**
   * Navigate from outside any screencomponent
   * @param routeName Name of ScreenComponent
   * @param params Any params to pass to the screen
   */
  navigate(routeName: string, params = {}) {
    console.info(this._navigator!.context)
    this._navigator!.dispatch(NavigationActions.navigate({ routeName, params }))
    // this._navigator!.dispatch(
    //   StackActions.reset({
    //     index: 0,
    //     actions: [NavigationActions.navigate({ routeName: 'Lidarr' })]
    //   })
    // )
  }

  private constructor() {}
}

export default RootNavigation.instance
