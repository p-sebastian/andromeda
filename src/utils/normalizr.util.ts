import { normalize, schema } from 'normalizr'
import {
  API_SONARR_GET_CALENDAR,
  API_SONARR_GET_SERIES,
  API_SONARR_GET_EPISODES
} from '@actions/types'
import { omit, groupBy } from 'lodash'
import { logger } from './logger.util'

// constant => AjaxResponse => normalizedRes
type Nrmlzr = (CONSTANT: string, json: any) => any
export const nrmlz: Nrmlzr = (CONSTANT, json) => {
  switch (CONSTANT) {
    case API_SONARR_GET_SERIES:
      return sonarrGetSeries(json)
    case API_SONARR_GET_CALENDAR:
      return sonarrGetCalendar(json)
    case API_SONARR_GET_EPISODES:
      return sonarrGetEpisodes(json)
    default:
      return json
  }
}

const sonarrGetSeries = (json: any) => {
  const series = new schema.Entity('series')
  const images = new schema.Entity(
    'images',
    {},
    {
      idAttribute: (entity, parent) => `${parent.id}-${entity.coverType}`
    }
  )
  series.define({
    images: [images]
  })
  const normal = normalize(json, [series])
  return normal
}

const sonarrGetCalendar = (json: any) => {
  const calendar = new schema.Entity(
    'calendar',
    {},
    {
      processStrategy: entity => omit(entity, 'series')
    }
  )
  return normalize(json, [calendar])
}

const sonarrGetEpisodes = (json: any) => {
  return groupBy(json, v => v.seasonNumber)
}
