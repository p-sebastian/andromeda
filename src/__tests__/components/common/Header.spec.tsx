import { BASE } from '@utils/theme.util'
import React from 'react'
import AHeader from '@common/Header.component'
import 'jest-styled-components/native'
import { withRedux } from '../../testing.utils'

describe('<AHeader/>', () => {
  it('renders', async () => {
    const { getByLabelText, asJSON } = withRedux(<AHeader />)
    expect(asJSON()).toMatchSnapshot()
    expect(getByLabelText(/header/)).toHaveStyleRule(
      'backgroundColor',
      BASE.primary
    )
  })
})
