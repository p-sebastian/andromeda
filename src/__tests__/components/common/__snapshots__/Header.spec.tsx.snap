// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AHeader/> renders 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      accessibilityLabel="header"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "hsla(225, 6%, 13%, 1)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
      theme={
        Object {
          "danger": "hsla(12.2, 81.9%, 54.8%, 1)",
          "dark": "hsla(225, 6%, 13%, 1)",
          "fontBold": "fira-code-bold",
          "fontColor": "hsla(225, 6%, 13%, 1)",
          "fontItalic": "dank-mono-italic",
          "fontRegular": "dank-mono",
          "lighterDark": "hsla(225, 9%, 16%, 1)",
          "primary": "hsla(225, 6%, 13%, 1)",
          "primaryDark": "hsla(232, 14%, 10%, 1)",
          "success": "hsla(133.6, 97%, 77.5%, 1)",
          "title": "Home",
        }
      }
    />
  </View>
</View>
`;
